devito/functions/dense.py

```diff
@@ -0,0 +1,1212 @@
+        right = [self._distributor.glb_to_loc(d, i, RIGHT, strict=False)
+                 for d, i in zip(self.dimensions, self._size_inhalo.right)]
+
+        Size = namedtuple('Size', 'left right')
+        sizes = tuple(Size(i, j) for i, j in zip(left, right))
+
+        return EnrichedTuple(*sizes, getters=self.dimensions, left=left, right=right)
+
+    @cached_property
+    def _mask_modulo(self):
+        """Boolean mask telling which Dimensions support modulo-indexing."""
+        return tuple(True if i.is_Stepping else False for i in self.dimensions)
+
+    @cached_property
+    def _mask_domain(self):
+        """Slice-based mask to access the domain region of the allocated data."""
+        return tuple(slice(i, j) for i, j in
+                     zip(self._offset_domain, self._offset_halo.right))
+
+    @cached_property
+    def _mask_inhalo(self):
+        """Slice-based mask to access the domain+inhalo region of the allocated data."""
+        return tuple(slice(i.left, i.right + j.right) for i, j in
+                     zip(self._offset_inhalo, self._size_inhalo))
+
+    @cached_property
+    def _mask_outhalo(self):
+        """Slice-based mask to access the domain+outhalo region of the allocated data."""
+        return tuple(slice(i.start - j.left, i.stop and i.stop + j.right or None)
+                     for i, j in zip(self._mask_domain, self._size_outhalo))
+
+    @cached_property
+    def _decomposition(self):
+        """
+        Tuple of Decomposition objects, representing the domain decomposition.
+        None is used as a placeholder for non-decomposed Dimensions.
+        """
+        if self._distributor is None:
+            return (None,)*self.ndim
+        mapper = {d: self._distributor.decomposition[d] for d in self._dist_dimensions}
+        return tuple(mapper.get(d) for d in self.dimensions)
+
+    @cached_property
+    def _decomposition_outhalo(self):
+        """
+        Tuple of Decomposition objects, representing the domain+outhalo
+        decomposition. None is used as a placeholder for non-decomposed Dimensions.
+        """
+        if self._distributor is None:
+            return (None,)*self.ndim
+        return tuple(v.reshape(*self._size_inhalo[d]) if v is not None else v
+                     for d, v in zip(self.dimensions, self._decomposition))
+
+    @property
+    def data(self):
+        """
+        The domain data values, as a numpy.ndarray.
+        Elements are stored in row-major format.
+        Notes
+        -----
+        With this accessor you are claiming that you will modify the values you
+        get back. If you only need to look at the values, use :meth:`data_ro`
+        instead.
+        """
+        return self.data_domain
+
+    @property
+    @_allocate_memory
+    def data_domain(self):
+        """
+        The domain data values.
+        Elements are stored in row-major format.
+        Notes
```

Comment:
You destroyed lots of docstrings in these files... need a blank line before notes
😭